### PR TITLE
Modify gathering process to be completely parallel

### LINKED ISSUE #55: Improve gathering process with parallelization
This is just to track the work related to making the gathering process completely parallel in order to improve the gathering speed.

### PR DESCRIPTION
**Pull request by trivigy**
resolves #55

### CODE REVIEW

**Comment by trivigy** [comment]
@masterada This is the parallelization you mentioned. Definitely improves the gathering speed.

File: gather.go [hunk 1]
```diff
@@ -146,48 +147,64 @@ func (a *Agent) gatherCandidates() {
...
+	var wg sync.WaitGroup
+	wg.Add(len(localIPs) * len(supportedNetworks))
 	for _, ip := range localIPs {
 		for _, network := range supportedNetworks {
-			conn, err := listenUDP(int(a.portmax), int(a.portmin), network, &net.UDPAddr{IP: ip, Port: 0})
-			if err != nil {
-				a.log.Warnf("could not listen %s %s\n", network, ip)
-				continue
-			}
+			go func(network string, ip net.IP) {
```
Comment (Sean-Der) [inline gather.go]: Possible to put the `wg.Done()` in a deferred?

Comment (trivigy) [reply gather.go]: done.

File: gather.go [hunk 2]
```diff
@@ -146,46 +147,62 @@ func (a *Agent) gatherCandidates() {
...
 
-			if err := a.run(func(agent *Agent) {
-				set := a.localCandidates[c.NetworkType()]
-				set = append(set, c)
-				a.localCandidates[c.NetworkType()] = set
-			}); err != nil {
-				a.log.Warnf("Failed to append to localCandidates: %v\n", err)
-				continue
-			}
+				if err := a.run(func(agent *Agent) {
```
Comment (masterada) [inline gather.go]: this is async, wg.Done can be called before the candidate is added, i suggest another wg (or maybe sync.Condition) here

**Comment by masterada** [comment]
nice :)

File: gather.go (hunk 2 above)
Comment (trivigy) [reply gather.go]: sync.Cond is likely to go away in go 2. I would recommend against that.
https://github.com/golang/go/issues/21165

Comment (trivigy) [reply gather.go]: p.s. `a.run` is not really async. It uses an unbuffered channel for the tasks which means only one task is running at a time. In this case, every `a.run` waits for it's turn to write into the channel. Only after that point it runs separately. What happens if you have another task running that is stalling?! Kinda challenging design with this `a.run` but it works.

Comment (trivigy) [reply gather.go]: @masterada If you still feel strongly about this specific comment let me know. I will try to figure out how to address it.

Comment (masterada) [reply gather.go]: nvm, you set `a.gatheringState = GatheringStateComplete` in a.run as well, so it definitely runs after all candidates are gathered (sorry, was looking on it on mobile :) )

File: gather.go [hunk 3]
```diff
@@ -146,46 +147,62 @@ func (a *Agent) gatherCandidates() {
...
-			conn, err := listenUDP(int(a.portmax), int(a.portmin), network, &net.UDPAddr{IP: ip, Port: 0})
-			if err != nil {
-				a.log.Warnf("could not listen %s %s\n", network, ip)
-				continue
-			}
+			go func(network string, ip net.IP) {
+				conn, err := listenUDP(int(a.portmax), int(a.portmin), network, &net.UDPAddr{IP: ip, Port: 0})
+				if err != nil {
+					a.log.Warnf("could not listen %s %s\n", network, ip)
+					wg.Done()
```
Comment (masterada) [inline gather.go]: defer wg.Done() (you call it before any return in this function)

Comment (trivigy) [reply gather.go]: done.

File: gather.go (hunk 1 above)
Comment (jinleileiking) [reply gather.go]: It seems defer wg.wait

**Pull request merged**